[Streams] View stream updates as HTTP requests

mykolaharmash opened the pull request:
Closes https://github.com/elastic/streams-program/issues/362

This change adds a way for users to view and/or copy an equivalent code with HTTP request information when making changes to a stream. The code later can be used to execute the requests manually or as part of infrastructure-as-code tooling (for example Terraform).

For all places where this feature is introduced, the code responsible for building the request's body was extracted into a helper function and reused for an actual request and for previewing the code.

## Retention editing modal
| "Copy code" button |
| --- |
| <img width="2818" height="1892" alt="CleanShot 2025-11-24 at 10 20 56@2x" src="https://github.com/user-attachments/assets/ff4743e3-2d27-41ce-a810-99d0f170051d" /> |

## Partitioning manual editor for creating a new child stream
| "View code" button | Flyout |
| --- | --- |
| <img width="2820" height="1786" alt="CleanShot 2025-11-24 at 10 21 50@2x" src="https://github.com/user-attachments/assets/ae6a1e57-5c53-4369-b2ca-7127d03727ec" /> | <img width="2818" height="2856" alt="CleanShot 2025-11-24 at 10 22 00@2x" src="https://github.com/user-attachments/assets/5437c01e-d330-4474-94a8-745f4e27fb76" /> |

## Partitioning manual editor for editing a new child stream
| "View code" button inside a context menu | "Copy code" button inside the delete confirmation modal |
| --- | --- |
|  <img width="2820" height="1736" alt="CleanShot 2025-11-24 at 16 23 19@2x" src="https://github.com/user-attachments/assets/5a2e51f9-c36e-46a0-a091-edc592c84a27" /> | <img width="2808" height="2404" alt="CleanShot 2025-11-24 at 16 23 37@2x" src="https://github.com/user-attachments/assets/0d54ca56-89d6-48fd-aafc-ca928708d13d" /> |

## Partitioning AI suggestions
| "Copy code" button in the confirmation modal |
| --- |
| <img width="2822" height="2458" alt="CleanShot 2025-11-24 at 10 29 39@2x" src="https://github.com/user-attachments/assets/72ce0fea-74e6-4d9b-b131-04e74ce79188" /> |

## Processing
| "View code" button inside the bottom bar | Flyout |
| --- | --- |
| <img width="2824" height="2956" alt="CleanShot 2025-11-24 at 10 32 19@2x" src="https://github.com/user-attachments/assets/3381b7b2-86cd-4fd6-951c-9b79cd5755c8" /> | <img width="2822" height="2954" alt="CleanShot 2025-11-24 at 10 32 32@2x" src="https://github.com/user-attachments/assets/0fdc1cba-1b7c-426d-9d5d-8c3b2e42564a" /> |

## Schema editor
| "View code" button inside the bottom bar | Flyout |
| --- | --- |
| <img width="2816" height="2854" alt="CleanShot 2025-11-24 at 10 33 06@2x" src="https://github.com/user-attachments/assets/048f1176-d957-4152-981f-c7bac4072e8b" /> | <img width="2820" height="2954" alt="CleanShot 2025-11-24 at 10 33 18@2x" src="https://github.com/user-attachments/assets/7a6bb67a-a351-4255-ab38-fc3402dde1ce" /> |

# How to test

1. You need a running inference service to test AI suggestions. [🔒 Login into Vault](https://docs.elastic.dev/vault/accessing), run `node scripts/eis.js` and follow the instructions to setup ES and Kibana.
2. Run Kibana and ES
3. Ingest some logs:
```bash
node scripts/synthtrace.js sample_logs --live --kibana=http://elastic:changeme@localhost:5601 --target=http://elastic:changeme@localhost:9200 --liveBucketSize=1000
```
4. Go through the places in Kibana that you see on the screenshots above and make sure the code visible in the flyout or copied to the clipboard is correct and corresponds to the actual requests we make.
5. Check that download button in the flyout works and the downloaded file has the correct code.

mykolaharmash committed:
[Streams] View stream updates as HTTP requests

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

mykolaharmash commented:
/ci

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/364811)
* Commit: 577229c57c2d5a94a2e8193b68f1c13e3d1d50a4

### Failed CI Steps
* [FTR Configs #82](https://buildkite.com/elastic/kibana-pull-request/builds/364811#019aba93-0536-44bb-af4a-24fb7e2084e5)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/364811#019aba93-0536-44bb-af4a-24fb7e2084e5) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/364811/jobs/019aba93-0536-44bb-af4a-24fb7e2084e5/artifacts/019abab7-a3c7-4210-8451-39668ee5f771) FTR Configs #<span></span>82 / Dataset Quality Dataset quality handles user privileges User can read logs-* User can monitor some data streams "View dashboards" is hidden for underprivileged user

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/72cad86aafd6abc23dac9c2e6abbdfcd34f0a90d) | [after](https://github.com/elastic/kibana/commit/577229c57c2d5a94a2e8193b68f1c13e3d1d50a4) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 1094 | 1096 | +2 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/72cad86aafd6abc23dac9c2e6abbdfcd34f0a90d) | [after](https://github.com/elastic/kibana/commit/577229c57c2d5a94a2e8193b68f1c13e3d1d50a4) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 1.0MB | 1.0MB | +7.1KB |



### History
* :broken_heart: [Build #364761](https://buildkite.com/elastic/kibana-pull-request/builds/364761) failed c940ae2c090ca4ddf0cbff95ad8d72fea1266484
* :broken_heart: [Build #364427](https://buildkite.com/elastic/kibana-pull-request/builds/364427) failed bfe52c673354d6a060661adfa7be806ddb8598e9
* :broken_heart: [Build #364196](https://buildkite.com/elastic/kibana-pull-request/builds/364196) failed 7d10d9c7b0ad75be4c6c59744549012038a2515a
* :broken_heart: [Build #364164](https://buildkite.com/elastic/kibana-pull-request/builds/364164) failed 9dd749475851966e76d85e8ebc08e93c684dd099

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364811","number":364811,"commit":"577229c57c2d5a94a2e8193b68f1c13e3d1d50a4"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364761","number":364761,"commit":"c940ae2c090ca4ddf0cbff95ad8d72fea1266484"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364427","number":364427,"commit":"bfe52c673354d6a060661adfa7be806ddb8598e9"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364196","number":364196,"commit":"7d10d9c7b0ad75be4c6c59744549012038a2515a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364164","number":364164,"commit":"9dd749475851966e76d85e8ebc08e93c684dd099"}],"number":364811}
buildkite-pr-comment-->

awahab07 commented on x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/page_content.tsx:
```diff
@@ -80,6 +88,7 @@ export function StreamDetailEnrichmentContentImpl() {
   const context = useKibana();
   const { appParams, core } = context;
 
+  const getStreamEnrichmentState = useGetStreamEnrichmentState();
```
Since `getStreamEnrichmentState` in only used for code flyout, does it make sense to have `useStreamEnrichmentCodeFlyout` instead which encapsulates the `isRequestPreviewFlyoutOpen`, `requestPreviewCodeContent`, `openRequestPreviewFlyout` and `closeRequestPreviewFlyout` and hides it from the component?

Or `useGetStreamEnrichmentState` exposes `{ ..., codeFlyout.*, ... }`?

awahab07 commented on x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_lifecycle/general_data/modal/modal.tsx:
```diff
@@ -116,6 +120,23 @@ export function EditLifecycleModal({
+    const body = buildLifecycleSaveRequestPayload(
+      definition,
+      isInheritToggleOn ? { inherit: {} } : lifecycle
+    );
```
nit: I think this logic which `"streamsAppModalFooterButton"` also uses, could be extracted into a reusable function in the same component. Then we may not need a separate function for `buildLifecycleSaveRequestPayload` as it's not doing much.

awahab07 reviewed (commented):
Code LGTM! Will test the functionality.

awahab07 reviewed (commented):
Tested the functionality and everything works. Thanks for the awesome implementation.

I've a few comments for the UI.

1. In "Partitioning" tab (routing view), since the "View code" button is per partition, it could be misleading if partitions above or below have changes as all the changes would be included in the request. See the vid:

    https://github.com/user-attachments/assets/4f912295-67d7-4ffe-8fdd-6e563784fb47

2. a. Also have some thoughts for UI, see:
  
 
    <img width="2280" height="1100" alt="image" src="https://github.com/user-attachments/assets/c8d85235-3775-4d75-be37-4f754471d393" />
 
  
     b. Similarly the button text in Retention is "Copy code", should it be "Copy API request"
    <img width="260" height="367" alt="image" src="https://github.com/user-attachments/assets/6272b1a0-e3a8-4490-b124-217a50dfffbf" />

mykolaharmash committed:
Merge remote-tracking branch 'upstream/main' into streams-http-export

mykolaharmash committed:
Add flyout helper hook and change copy

mykolaharmash commented:
Thank you for the suggestions @awahab07! I pushed an updated version, please take a look.

Summary of the changes mostly following your comments:
* Added a helper hook that handles flyout state
* Added a helper function that handle the updated lifecycle object and got rid of `buildLifecycleSaveRequestPayload()`
* Changed "View code" to "View API request" everywhere
* Changed copy in the flyout: "Cancel" → "Close" and the description above the code snippet. I've decided not split the URL and the code to keep it consistent with the Copy and Download buttons that give user the full snippet. Plus it also plays better with the global "Copy to clipboard" button in the flyout which I think is better to keep as is gives a clear CTA to the flyout.
* Updated the button layout in the child stream editor according to the latest designs

awahab07 reviewed (approved):
Tested again and it LGTM! Thanks for implementing this.

The "update routing per card issue", it's not specific to "View API request" dialog, but also how the routing UI currently behaves. 

@patpascal @boriskirov see the following video where I've demonstrated this behavior, and see if it's a UX concern:

The per card **Update** button would save routing definition for all children.

https://github.com/user-attachments/assets/49ef7396-60b3-46fa-b4be-04273575507d

_also see [related](https://github.com/elastic/kibana/pull/242674#pullrequestreview-3516125659)_ 

Though the per card **Remove** and **Save** are good.

https://github.com/user-attachments/assets/cc9f41a6-e617-4762-aee8-7a56da3fcac8

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/366773)
* Commit: 3ed7d599e72452aeb1fc50164cfd0a59403d4a5f

### Failed CI Steps
* [FTR Configs #15](https://buildkite.com/elastic/kibana-pull-request/builds/366773#019acaf4-17b8-4e09-8fca-01eda9717791)
* [FTR Configs #15](https://buildkite.com/elastic/kibana-pull-request/builds/366773#019acb17-768a-4009-8d52-0ef855f7d418)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/366773#019acaf4-17b8-4e09-8fca-01eda9717791) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/366773/jobs/019acaf4-17b8-4e09-8fca-01eda9717791/artifacts/019acb17-5dee-4b4c-9ba2-181172a3c507) FTR Configs #<span></span>15 / ObservabilityApp Custom threshold rule with ad-hoc data view should create a custom threshold rule with an ad-hoc data view
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/366773#019acb17-768a-4009-8d52-0ef855f7d418) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/366773/jobs/019acb17-768a-4009-8d52-0ef855f7d418/artifacts/019acb27-95b6-41a1-b490-8eab485cd534) FTR Configs #<span></span>15 / ObservabilityApp Custom threshold rule with ad-hoc data view should create a custom threshold rule with an ad-hoc data view

### History
* :yellow_heart: [Build #364811](https://buildkite.com/elastic/kibana-pull-request/builds/364811) was flaky 577229c57c2d5a94a2e8193b68f1c13e3d1d50a4
* :broken_heart: [Build #364761](https://buildkite.com/elastic/kibana-pull-request/builds/364761) failed c940ae2c090ca4ddf0cbff95ad8d72fea1266484
* :broken_heart: [Build #364427](https://buildkite.com/elastic/kibana-pull-request/builds/364427) failed bfe52c673354d6a060661adfa7be806ddb8598e9
* :broken_heart: [Build #364196](https://buildkite.com/elastic/kibana-pull-request/builds/364196) failed 7d10d9c7b0ad75be4c6c59744549012038a2515a
* :broken_heart: [Build #364164](https://buildkite.com/elastic/kibana-pull-request/builds/364164) failed 9dd749475851966e76d85e8ebc08e93c684dd099

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366773","number":366773,"commit":"3ed7d599e72452aeb1fc50164cfd0a59403d4a5f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364811","number":364811,"commit":"577229c57c2d5a94a2e8193b68f1c13e3d1d50a4"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364761","number":364761,"commit":"c940ae2c090ca4ddf0cbff95ad8d72fea1266484"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364427","number":364427,"commit":"bfe52c673354d6a060661adfa7be806ddb8598e9"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364196","number":364196,"commit":"7d10d9c7b0ad75be4c6c59744549012038a2515a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364164","number":364164,"commit":"9dd749475851966e76d85e8ebc08e93c684dd099"}],"number":366773}
buildkite-pr-comment-->

mykolaharmash committed:
Merge remote-tracking branch 'upstream/main' into streams-http-export
